mlos_bench/mlos_bench/optimizers/grid_search_optimizer.py

```diff
@@ -0,0 +1,190 @@
+        """
+        Generate the next grid search suggestion.
+        """
+        tunables = self._tunables.copy()
+        if self._start_with_defaults:
+            _LOG.info("Use default values for the first trial")
+            self._start_with_defaults = False
+            tunables = tunables.restore_defaults()
+            # Need to index based on ConfigSpace dict ordering.
+            default_config = dict(self.config_space.get_default_configuration())
+            assert tunables.get_param_values() == default_config
+            # Move the default from the pending to the suggested set.
+            default_config_values = tuple(default_config.values())
+            del self._pending_configs[default_config_values]
+            self._suggested_configs.add(default_config_values)
+        else:
+            # Select the first item from the pending configs.
+            if not self._pending_configs and self._iter <= self._max_iter:
+                _LOG.info("No more pending configs to suggest. Restarting grid.")
+                self._config_keys, self._pending_configs = self._get_grid()
+            try:
+                next_config_values = next(iter(self._pending_configs.keys()))
+            except StopIteration as exc:
+                raise ValueError("No more pending configs to suggest.") from exc
+            next_config = dict(zip(self._config_keys, next_config_values))
+            tunables.assign(next_config)
+            # Move it to the suggested set.
+            self._suggested_configs.add(next_config_values)
+            del self._pending_configs[next_config_values]
+        _LOG.info("Iteration %d :: Suggest: %s", self._iter, tunables)
+        return tunables
+
+    def register(self, tunables: TunableGroups, status: Status,
+                 score: Optional[Union[float, dict]] = None) -> Optional[float]:
+        registered_score = super().register(tunables, status, score)
+        if status.is_succeeded() and (
+            self._best_score is None or (registered_score is not None and registered_score < self._best_score)
+        ):
+            self._best_score = registered_score
+            self._best_config = tunables.copy()
+        self._iter += 1
+        try:
+            config = dict(ConfigSpace.Configuration(self.config_space, values=tunables.get_param_values()))
+            self._suggested_configs.remove(tuple(config.values()))
+        except KeyError:
+            _LOG.warning("Attempted to remove missing config (previously registered?) from suggested set: %s", tunables)
+        return registered_score
+
+    def not_converged(self) -> bool:
+        if self._iter > self._max_iter:
+            if bool(self._pending_configs):
+                _LOG.warning("Exceeded max iterations, but still have %d pending configs: %s",
+                             len(self._pending_configs), list(self._pending_configs.keys()))
+        return self._iter <= self._max_iter and bool(self._pending_configs)
```

Comment:
maybe
```suggestion
        if self._iter > self._max_iter:
            if bool(self._pending_configs):
                _LOG.warning("Exceeded max iterations, but still have %d pending configs: %s",
                             len(self._pending_configs), list(self._pending_configs.keys()))
            return False
        return bool(self._pending_configs)
```
say, if I want to run only 5 iterations of the larger grid search

BTW, why `bool()`? Is that something mypy wants?

Reply:
re `bool()` yeah I think it was a type thing, but I can double check.

The other logic I think covers your case as well btw - it's even checked in the tests, but it may not be as readable.

Should evaluate to `return False and bool(self._pending_configs) => False`